File: core/contractsdk/cpp/example/award_manage.cc
#include "xchain/xchain.h"

const std::string BALANCEPRE = "balanceOf_";
const std::string ALLOWANCEPRE = "allowanceOf_";
const std::string MASTERPRE = "owner";

// 积分管理合约的基类
// 积分管理合约需要实现基类中指定的方法
// 参数由xchain::Contract中的context提供
class AwardBasic {
public:
    /*
     * func: 初始化积分管理账户以及总发行量
     * @param: initiator:交易发起者,也是初始化积分的owner
     * @param: totalSupply:发行总量,初始化时,积分全部归initiator
     */
    virtual void initialize() = 0;
    /*
     * func: 增发积分
     * @param: initiator:交易发起者,只有交易发起者等于积分owner时，才能增发
     * @param: amount:增发容量
     */
    virtual void addAward() = 0;
    /*
     * func: 获取积分总供应量
     */
    virtual void totalSupply() = 0;
    /*
     * func: 获取caller的积分余额
     * @param: caller: 合约调用者
     */
    virtual void balance() = 0;
    /*
     * func: 查询to用户能消费from用户的积分数量
     * @param: from: 被消费积分的一方
     * @param: to: 消费积分的一方
     */
    virtual void allowance() = 0;
    /*
     * func: from账户给to账户转token数量的积分
     * @param: from:转移积分的一方
     * @param: to:收积分的一方
     * @param: token:转移积分数量
     */
    virtual void transfer() = 0;
    /*
     * func: 从授权账户from转移数量为token的积分给to账户
      * @param: from:被转积分账户
      * @param: caller:合约调用者
      * @param: to:收积分账户
      * @param: token:转移的积分数量
      */
    virtual void transferFrom() = 0;
    /*
      * func: 允许to账户从from账户转移token数量的积分
     * @param: from:
     * @param: to:
     * @param: token
     */
    virtual void approve() = 0;
};

struct Award : public AwardBasic, public xchain::Contract {
public:
    void initialize() {
        xchain::Context* ctx = this->context();
        const std::string& caller = ctx->initiator();
        if (caller.empty()) {
            ctx->error("missing caller");
            return;
        }
        const std::string& totalSupply = ctx->arg("totalSupply");
        if (totalSupply.empty()) {
            ctx->error("missing totalSupply");
            return;
        }
        if (atoi(totalSupply.c_str()) <= 0) {
            ctx->error("totalSupply is overflow");
            return;
        }

        std::string key = BALANCEPRE + caller;
        ctx->put_object("totalSupply", totalSupply);
        ctx->put_object(key, totalSupply);

        std::string master = MASTERPRE;
        ctx->put_object(master, caller);
        ctx->ok("initialize success");
    }
    void addAward() {
        xchain::Context* ctx = this->context();
        const std::string& caller = ctx->initiator();
        if (caller.empty()) {
            ctx->error("missing caller");
            return;
        }

        std::string master;
        if (!ctx->get_object(MASTERPRE, &master)) {
            ctx->error("missing master");
            return;
        }
        if (master != caller) {
            ctx->error("only the person who created the contract can addAward");
            return;
        }

        const std::string& increaseSupply = ctx->arg("amount");
        if (increaseSupply.empty()) {
            ctx->error("missing amount");
            return;
        }
        if (atoi(increaseSupply.c_str()) <= 0) {
            ctx->error("amount is overflow");
            return;
        }

        std::string value;
        if (!ctx->get_object("totalSupply", &value)) {
            ctx->error("get totalSupply error");
            return;
        }
        int increaseSupplyint = atoi(increaseSupply.c_str());
        int valueint = atoi(value.c_str());
        int totalSupplyint = increaseSupplyint + valueint;
        if (totalSupplyint <= 0) {
            ctx->error("amount+totalSupply is overflow");
            return;
        }
        char buf[32];
        snprintf(buf, 32, "%d", totalSupplyint);
        ctx->put_object("totalSupply", buf); 

        std::string key = BALANCEPRE + caller;
        if (!ctx->get_object(key, &value)) {
            ctx->error("get caller balance error");
            return;
        }
        valueint = atoi(value.c_str());
        int callerint = increaseSupplyint + valueint;
        snprintf(buf, 32, "%d", callerint);
        ctx->put_object(key, buf); 
    
        ctx->ok(buf);
    }
    void totalSupply() {
        xchain::Context* ctx = this->context();
        std::string value;
        if (ctx->get_object("totalSupply", &value)) {
            ctx->ok(value);
        } else {
            ctx->error("key not found");
        }
    }
    void balance() {
        xchain::Context* ctx = this->context();
        const std::string& caller = ctx->arg("caller");
        if (caller.empty()) {
            ctx->error("missing caller");
            return;
        }

        std::string key = BALANCEPRE + caller;
        std::string value;
        if (ctx->get_object(key, &value)) {
            ctx->ok(value);
        } else {
            ctx->error("key not found");
        }
    }
    void allowance() {
        xchain::Context* ctx = this->context();
        const std::string& from = ctx->arg("from");
        if (from.empty()) {
            ctx->error("missing from");
            return;
        }

        const std::string& to = ctx->arg("to");
        if (to.empty()) {
            ctx->error("missing to");
            return;
        }

        std::string key = ALLOWANCEPRE + from + "_" + to;
        std::string value;
        if (ctx->get_object(key, &value)) {
            ctx->ok(value);
        } else {
            ctx->error("key not found");
        }
    }
    void transfer() {
        xchain::Context* ctx = this->context();
        const std::string& from = ctx->arg("from");
        if (from.empty()) {
            ctx->error("missing from");
            return;
        }
   
        const std::string& to = ctx->arg("to");
        if (to.empty()) {
            ctx->error("missing to");
            return;
        }

        const std::string& token_str = ctx->arg("token");
        if (token_str.empty()) {
            ctx->error("missing token");
            return;
        }
        int token = atoi(token_str.c_str());
        if (token <= 0) {
            ctx->error("token is overflow");
            return;
        }

        std::string from_key = BALANCEPRE + from;
        std::string value;
        int from_balance = 0;
        if (ctx->get_object(from_key, &value)) {
            from_balance = atoi(value.c_str()); 
            if (from_balance < token) {
                ctx->error("The balance of from not enough");
                return;
            }  
        } else {
            ctx->error("key not found");
            return;
        }

        std::string to_key = BALANCEPRE + to;
        int to_balance = 0;
        if (ctx->get_object(to_key, &value)) {
            to_balance = atoi(value.c_str());
        }
   
        from_balance = from_balance - token;
        to_balance = to_balance + token;
   
        char buf[32]; 
        snprintf(buf, 32, "%d", from_balance);
        ctx->put_object(from_key, buf);
        snprintf(buf, 32, "%d", to_balance);
        ctx->put_object(to_key, buf);

        ctx->ok("transfer success");
    }
    void transferFrom() {
        xchain::Context* ctx = this->context();
        const std::string& from = ctx->arg("from");
        if (from.empty()) {
            ctx->error("missing from");
            return;
        }
  
        const std::string& caller = ctx->arg("caller");
        if (caller.empty()) {
            ctx->error("missing caller");
            return;
        }

        const std::string& to = ctx->arg("to");
        if (to.empty()) {
            ctx->error("missing to");
            return;
        }

        const std::string& token_str = ctx->arg("token");
        if (token_str.empty()) {
            ctx->error("missing token");
            return;
        }
        int token = atoi(token_str.c_str());
        if (token <= 0) {
            ctx->error("token is overflow");
            return;
        }

        std::string allowance_key = ALLOWANCEPRE + from + "_" + caller;
        std::string value;
        int allowance_balance = 0;
        if (ctx->get_object(allowance_key, &value)) {
            allowance_balance = atoi(value.c_str()); 
            if (allowance_balance < token) {
                ctx->error("The allowance of from_to not enough");
                return;
            }  
        } else {
            ctx->error("You need to add allowance from_to");
            return;
        }

        std::string from_key = BALANCEPRE + from;
        int from_balance = 0;
        if (ctx->get_object(from_key, &value)) {
            from_balance = atoi(value.c_str()); 
            if (from_balance < token) {
                ctx->error("The balance of from not enough");
                return;
            }  
        } else {
            ctx->error("From no balance");
            return;
        }

        std::string to_key = BALANCEPRE + to;
        int to_balance = 0;
        if (ctx->get_object(to_key, &value)) {
            to_balance = atoi(value.c_str());
        }
   
        from_balance = from_balance - token;
        to_balance = to_balance + token;
        allowance_balance = allowance_balance - token;

        char buf[32]; 
        snprintf(buf, 32, "%d", from_balance);
        ctx->put_object(from_key, buf);
        snprintf(buf, 32, "%d", to_balance);
        ctx->put_object(to_key, buf);
        snprintf(buf, 32, "%d", allowance_balance);
        ctx->put_object(allowance_key, buf);

        ctx->ok("transferFrom success");
    }
    void approve() {
        xchain::Context* ctx = this->context();
        const std::string& from = ctx->arg("from");
        if (from.empty()) {
            ctx->error("missing from");
            return;
        }
   
        const std::string& to = ctx->arg("to");
        if (to.empty()) {
            ctx->error("missing to");
            return;
        }

        const std::string& token_str = ctx->arg("token");
        if (token_str.empty()) {
            ctx->error("missing token");
            return;
        }
        int token = atoi(token_str.c_str());
        if (token <= 0) {
            ctx->error("token is overflow");
            return;
        }

        std::string from_key = BALANCEPRE + from;
        std::string value;
        if (ctx->get_object(from_key, &value)) {
            int from_balance = atoi(value.c_str()); 
            if (from_balance < token) {
                ctx->error("The balance of from not enough");
                return;
            }  
        } else {
            ctx->error("From no balance");
            return;
        }

        std::string allowance_key = ALLOWANCEPRE + from + "_" + to;
        int allowance_balance = 0;
        if (ctx->get_object(allowance_key, &value)) {
            allowance_balance = atoi(value.c_str());
        }

        allowance_balance = allowance_balance + token;
   
        char buf[32]; 
        snprintf(buf, 32, "%d", allowance_balance);
        ctx->put_object(allowance_key, buf);

        ctx->ok("approve success");
    }
};


DEFINE_METHOD(Award, initialize) {
    self.initialize();
}

DEFINE_METHOD(Award, addAward) {
    self.addAward();
}

DEFINE_METHOD(Award, totalSupply) {
    self.totalSupply();
}

DEFINE_METHOD(Award, balance) {
    self.balance();
}

DEFINE_METHOD(Award, allowance) {
    self.allowance();
}

DEFINE_METHOD(Award, transfer) {
    self.transfer();
}

DEFINE_METHOD(Award, transferFrom) {
    self.transferFrom();
}

DEFINE_METHOD(Award, approve) {
    self.approve();
}

File: core/contractsdk/java/example/award-manage/src/main/java/com/baidu/xuper/AwardManage.java
package com.baidu.xuper;

import java.math.BigInteger;

import com.baidu.xuper.Context;
import com.baidu.xuper.Contract;
import com.baidu.xuper.ContractMethod;
import com.baidu.xuper.Driver;
import com.baidu.xuper.Response;

public class AwardManage implements Contract {
    final String BALANCE = "balanceOf_";
    final String ALLOWANCE = "allowanceOf_";
    final String MASTER = "owner";

    @Override
    @ContractMethod
    public Response initialize(Context ctx) {
        String caller = ctx.caller();
        if (caller.length() == 0) {
            return Response.error("missing caller");
        }
        byte[] totalSupplyByte = ctx.args().get("totalSupply");
        if (totalSupplyByte == null || totalSupplyByte.length == 0) {
            return Response.error("missing totalSupply");
        }
        BigInteger totalSupply = new BigInteger(new String(totalSupplyByte));
        if (totalSupply.compareTo(BigInteger.ZERO) <= 0) {
            return Response.error("totalSupply must be positive");
        }

        ctx.putObject((BALANCE + caller).getBytes(), totalSupply.toString().getBytes());
        ctx.putObject("totalSupply".getBytes(), totalSupply.toString().getBytes());
        ctx.putObject(MASTER.getBytes(), caller.getBytes());

        return Response.ok("ok".getBytes());
    }

    private boolean permCheck(Context ctx) {
        return ctx.caller() != null && !ctx.caller().isEmpty()
                && ctx.caller().equals(new String(ctx.getObject(MASTER.getBytes())));
    }

    @ContractMethod
    public Response addAward(Context ctx) {
        if (!permCheck(ctx)) {
            return Response.error("you do not have permission to call this method");
        }
        byte[] amountByte = ctx.args().get("amount");
        if (amountByte == null || amountByte.length == 0) {
            return Response.error("missing amount");
        }

        BigInteger amount = new BigInteger(new String(ctx.args().get("amount")));
        if (amount.compareTo(BigInteger.ZERO) <= 0) {
            return Response.error("amount must be greater than 0");
        }

        BigInteger totalSupply = new BigInteger(new String(ctx.getObject("totalSupply".getBytes())));

        ctx.putObject("totalSupply".getBytes(), totalSupply.add(amount).toString().getBytes());

        String key = BALANCE + ctx.caller();
        BigInteger value = new BigInteger(new String(ctx.getObject(key.getBytes())));

        ctx.putObject(key.getBytes(), value.add(amount).toString().getBytes());
        return Response.ok("ok~".getBytes());
    }

    @ContractMethod
    public Response totalSupply(Context ctx) {
        return Response.ok(ctx.getObject("totalSupply".getBytes()));
    }

    @ContractMethod
    public Response balance(Context ctx) {
        String caller = ctx.caller();
        if (caller == null || caller.length() == 0) {
            return Response.error("missing caller");
        }
        byte[] value = ctx.getObject((BALANCE + caller).getBytes());
        return Response.ok(value);
    }

    @ContractMethod
    public Response allowance(Context ctx) {
        byte[] fromByte = ctx.args().get("from");
        byte[] toByte = ctx.args().get("to");
        if (fromByte == null || fromByte.length == 0 || toByte == null || toByte.length == 0) {
            return Response.error("missing from or to");
        }
        String from = new String(fromByte);
        String to = new String(toByte);
        byte[] allowanceByte = ctx.getObject((ALLOWANCE + from + "_" + to).getBytes());
        if (allowanceByte == null || allowanceByte.length == 0) {
            return Response.error("allowance from " + from + " to " + to + " not found");
        }
        return Response.ok(allowanceByte);

    }

    @ContractMethod
    public Response transfer(Context ctx) {
        String from = ctx.caller();
        if (from == null || from.isEmpty()) {
            return Response.error("missing caller");
        }
        byte[] toByte = ctx.args().get("to");
        byte[] tokenByte = ctx.args().get("token");
        if (toByte == null || toByte.length == 0 || tokenByte == null || toByte.length == 0) {
            return Response.error("missing  to or token");
        }

        String to = new String((toByte));
        BigInteger token = new BigInteger(new String(tokenByte));

        if (from.equals(to)) {
            return Response.error("can not transfer to yourself");
        }

        if (token.compareTo(BigInteger.ZERO) <= 0) {
            return Response.error("token must be more than 0");
        }

        String fromKey = BALANCE + from;
        byte[] balanceFromByte = ctx.getObject(fromKey.getBytes());
        if (balanceFromByte == null || balanceFromByte.length == 0) {
            return Response.error("balance of " + from + " not found");
        }
        BigInteger fromBalance = new BigInteger(new String(balanceFromByte));
        if (fromBalance.compareTo(token) < 0) {
            return Response.error("balance not enough");
        }

        String toKey = BALANCE + to;
        byte[] balanceToByte = ctx.getObject(toKey.getBytes());
        BigInteger toBalance = new BigInteger("0");
        if (balanceToByte != null && balanceToByte.length != 0) {
            toBalance = new BigInteger(new String(balanceToByte));
        }

        ctx.putObject(fromKey.getBytes(), fromBalance.subtract(token).toString().getBytes());
        ctx.putObject(toKey.getBytes(), toBalance.add(token).toString().getBytes());
        return Response.ok("ok~".getBytes());
    }

    @ContractMethod
    public Response transferFrom(Context ctx) {
        String to = ctx.caller();
        if (to == null || to.isEmpty()) {
            return Response.error("missing caller");
        }

        if (ctx.args().get("from") == null || ctx.args().get("from").length == 0 || ctx.args().get("token") == null
                || ctx.args().get("token").length == 0) {
            return Response.error("missing from or token");
        }

        String from = new String(ctx.args().get("from"));
        BigInteger token = new BigInteger(new String(ctx.args().get("token")));

        if (token.compareTo(BigInteger.ZERO) == 0) {
            return Response.error("missing token");
        }
        if (from.equals(to)) {
            return Response.error("can not transfer from yourself");
        }
        if (token.compareTo(BigInteger.ZERO) <= 0) {
            return Response.error("token must be more than 0");
        }

        String allowanceKey = ALLOWANCE + from + "_" + to;
        byte[] allowanceByte = ctx.getObject(allowanceKey.getBytes());
        if (allowanceByte == null || allowanceByte.length == 0) {
            return Response.error("allowance from " + from + " to " + to + " not found");
        }

        BigInteger allowance = new BigInteger(new String(allowanceByte));

        if (allowance.compareTo(token) < 0) {
            return Response.error("allowance not enough");
        }
        String fromKey = BALANCE + from;
        String toKey = BALANCE + to;
        byte[] fromBalanceByte = ctx.getObject(fromKey.getBytes());
        if (fromBalanceByte == null || fromBalanceByte.length == 0) {
            return Response.error("balance of" + from + " not found");
        }

        BigInteger fromBalance = new BigInteger(new String(fromBalanceByte));
        if (fromBalance.compareTo(token) < 0) {
            return Response.error("from balancfe not enough");
        }

        BigInteger toBalance = new BigInteger("0");
        byte[] value = ctx.getObject((BALANCE + to).getBytes());
        if (value != null && value.length > 0) {
            toBalance = new BigInteger(new String(value));
        }
        ctx.putObject(fromKey.getBytes(), fromBalance.subtract(token).toString().getBytes());
        ctx.putObject((toKey).getBytes(), toBalance.add(token).toString().getBytes());
        ctx.putObject(allowanceKey.getBytes(), allowance.subtract(token).toString().getBytes());
        return Response.ok("ok~".getBytes());
    }

    @ContractMethod
    public Response approve(Context ctx) {
        String from = ctx.caller();
        if (from == null || from.isEmpty()) {
            return Response.error("missing caller");
        }

        if (ctx.caller() == null || ctx.caller().isEmpty() || ctx.args().get("token") == null
                || ctx.args().get("token").length == 0 || ctx.args().get("to") == null
                || ctx.args().get("to").length == 0) {
            return Response.error("missing caller to or token");
        }

        String to = new String(ctx.args().get("to"));
        BigInteger token = new BigInteger(new String(ctx.args().get("token")));

        if (token.compareTo(BigInteger.ZERO) <= 0) {
            return Response.error("token must be greater than 0");
        }
        if (from.equals(to)) {
            return Response.error("can not transfer to yourself");
        }

        String allowanceKey = ALLOWANCE + from + "_" + to;
        BigInteger allowance = new BigInteger("0");
        byte[] value = ctx.getObject(allowanceKey.getBytes());
        if (value != null) {
            allowance = new BigInteger(new String(value));
        }

        allowance = allowance.add(token);
        ctx.putObject(allowanceKey.getBytes(), allowance.toString().getBytes());
        return Response.ok("ok~".getBytes());

    }

    public static void main(String[] args) {
        Driver.serve(new AwardManage());
    }
}